controllers/component_dependency_update_controller.go

```diff
@@ -167,31 +175,29 @@ func (r *ComponentDependencyUpdateReconciler) Reconcile(ctx context.Context, req
 	log.Info("reconciling PipelineRun")
 
 	if pipelineRun.IsDone() || pipelineRun.Status.CompletionTime != nil || pipelineRun.DeletionTimestamp != nil {
-		log.Info("PipelineRun complete")
-		// These objects are so heavily contented that we always grab the latest copy from the
-		// API server to reduce the chance of conflicts
-		err = r.ApiReader.Get(ctx, req.NamespacedName, pipelineRun)
-		if err != nil {
-			return ctrl.Result{}, err
+		result, err := r.verifyUpToDate(ctx, pipelineRun)
+		if err != nil || result != nil {
+			return *result, err
 		}
+		log.Info("PipelineRun complete")
 		if controllerutil.ContainsFinalizer(pipelineRun, NudgeFinalizer) || pipelineRun.Annotations == nil || pipelineRun.Annotations[NudgeProcessedAnnotationName] == "" {
 			log.Info("running renovate job")
 			// Pipeline run is done and we have not cleared the finalizer yet
 			// We need to perform our nudge
-			return r.handleCompletedBuild(ctx, pipelineRun, component)
+			patch := client.MergeFrom(pipelineRun.DeepCopy())
+			return r.handleCompletedBuild(ctx, pipelineRun, component, patch)
 		}
 	} else if !controllerutil.ContainsFinalizer(pipelineRun, NudgeFinalizer) {
-		log.Info("adding finalizer for component nudge")
-		// These objects are so heavily contented that we always grab the latest copy from the
-		// API server to reduce the chance of conflicts
-		err := r.ApiReader.Get(ctx, req.NamespacedName, pipelineRun)
-		if err != nil {
-			return ctrl.Result{}, err
+		result, err := r.verifyUpToDate(ctx, pipelineRun)
+		if err != nil || result != nil {
+			return *result, err
 		}
+		log.Info("adding finalizer for component nudge")
 		// We add a finalizer to make sure we see the run before it is deleted
 		// As tekton results should aggressivly delete when pruning is enabled
+		patch := client.MergeFrom(pipelineRun.DeepCopy())
 		controllerutil.AddFinalizer(pipelineRun, NudgeFinalizer)
-		err = r.Client.Update(ctx, pipelineRun)
+		err = r.Client.Patch(ctx, pipelineRun, patch)
```

Comment:
If I'm reading it right, the `patch` is a deepcopy of the original pipelineRun object, without the finalizer. And then the code adds the finalizer to the `pipelineRun` object, and then calls `Patch()`. How does that work? I would expect the finalizer to be on the patch, not on the object we're patching

Reply:
Ohh, it's flipped for MergePatch https://github.com/kubernetes-sigs/controller-runtime/blob/1b80b9629af8ec55ae200e2c649a9603533eed20/pkg/client/patch.go#L100-L102